### PR TITLE
feat: translate date and time according to user language preference in mails & invoice

### LINKED ISSUE #8337: Change language of date and time based on user choice
The language for date and time is not changed in messages even if the message is not in English.

### PR DESCRIPTION
**Pull request by progmatic-99**
Fixes #8337

### CODE REVIEW

**Comment by mariobehling** [comment]
Thanks. Could you please look into the DeepSource test that fails?

**Comment by progmatic-99** [comment]
@codedsun @pc-beast please check on your local machines

File: app/api/helpers/mail.py [hunk 1]
```diff
@@ -794,11 +822,23 @@ def send_email_after_event_speaker(email, event_name):
     )
 
 
-def convert_to_event_timezone(date, timezone=None, format='%B %d, %Y %H:%M (%Z%z)'):
-    if not date:
+def convert_to_event_timezone(date, email, timezone=None, fmt='full'):
+    if not date or not email:
```
Comment (maze-runnar) [inline app/api/helpers/mail.py:826]: what if `if date and not email`?

**Review by maze-runnar — CHANGES_REQUESTED** [review]

File: app/api/helpers/mail.py [hunk 2]
```diff
@@ -794,11 +822,26 @@ def send_email_after_event_speaker(email, event_name):
...
+
+def convert_to_user_locale(email, date, fmt):
+    """Convert datetime to user selected language"""
+    if date and email:
+        user = User.query.filter(User.email == email).first()
+        user_locale = 'en'
+
+        if user and user.language_prefrence:
+            user_locale = user.language_prefrence
+            return format_datetime(date, format=fmt, locale=user_locale)
```
Comment (pc-beast) [inline app/api/helpers/mail.py:843]: This line can be removed.

File: app/api/helpers/mail.py [hunk 3]
```diff
@@ -554,13 +561,6 @@ def send_email_to_attendees(order):
         order_view_url=order.site_view_link,
     )
 
-    event_date = convert_to_event_timezone(
-        order.event.starts_at, order.event.timezone, format='%d %B %Y'
-    )
-    event_time = convert_to_event_timezone(
-        order.event.starts_at, order.event.timezone, format='%H:%M (%Z%z)'
-    )
-
```
Comment (pc-beast) [inline app/api/helpers/mail.py:563]: Why remove these? Isn't it better to just change the parameters as this is being used repetitively. :thinking:

**Review by pc-beast — CHANGES_REQUESTED** [review]

**Comment by progmatic-99** [comment]
**ticket_purchased_organizer email:**
ss_1637348080

**ticket_purchased** email
ss_1637348113
ss_1637348128

**ticket_attendee** pdf
ss_1637348140

**Comment by mariobehling** [comment]
Thanks. I am confused about the example emails. I see the text is in English and the date/time is in German here. As we already translated nearly everything to German I would expect everything to be in German in this example.

**Comment by progmatic-99** [comment]
@mariobehling My browser language is set to English, that's why this happened.

**ticket_attendee** mail
ss_1637385702

**ticket_purchased_organizer** mail
ss_1637385734

**Review by mariobehling — APPROVED** [review]
Looks good for me. Would be good if @maze-runnar could review the changes.

**Comment by maze-runnar** [comment]
Code changes are ok, but repeated a lot. If we have to convert startsat and endsat so many places, then directly convert them in schema at only one place.

**Comment by progmatic-99** [comment]
@maze-runnar In some mails, we need the date part of datetime object, in others we want only time part and in some places we want datetime + timezone info, so for which of the 3 cases I will convert in the schema??

**Comment by mariobehling** [comment]
@progmatic-99 Could you please change the PR title/commit message to something a bit more meaningful as well? Thank you.

**Comment by maze-runnar** [comment]
> @maze-runnar In some mails, we need the date part of datetime object, in others we want only time part and in some places we want datetime + timezone info, so for which of the 3 cases I will convert in the schema??

Okay then then this method is ok. We can merge this, and then think of a way to improve it

**Comment by mariobehling** [comment]
Regarding this
> My browser language is set to English, that's why this happened.

No matter what language your browser is set it will still show in the original language. Only if you activated auto-translate, e.g. in Chrome, then it will show in English or to whatever you have set it. The subject language has nothing to do with the browser. The email is sent out by the server and needs to be solved in the server.

**Comment by progmatic-99** [comment]
There is a different issue #8333 for subjects of mails. This can be merged @mariobehling @pc-beast @codedsun

**Review by pc-beast — APPROVED** [review]

**Comment by pc-beast** [comment]
@mariobehling can be merged now

**Review by codedsun — APPROVED** [review]

**Pull request merged**